### PR TITLE
Upgrade dependencies 2023-12-11 (#5772)

### LINKED ISSUE #5772: Upgrade dependencies 2023-12-11
- [x] Update [PyCharm image](https://github.com/DataBiosphere/azul-docker-pycharm)
  - [x] Bump [base image](https://hub.docker.com/_/debian/tags?name=bullseye) tag (only same Debian release), if possible
  - [x] Bump upstream version, if possible
  - [x] Bump internal version
  - [x] Build and test new image locally with Azul's `make format`
  - [x] Remove unused dependencies with high or critical CVEs
  - [x] Push commit to GitHub (directly to master branch, no PR needed)
  - [x] GH Action workflow succeeded
  - [x] Image is available on [DockerHub](https://hub.docker.com/repository/docker/ucscgi/azul-pycharm) 
- [x] Update [Elasticsearch image](https://github.com/DataBiosphere/azul-docker-elasticsearch)
  - [x] Bump [base image](https://hub.docker.com/_/elasticsearch/tags) tag (only minor and patch versions), if possible
  - [x] Bump internal version 
  - [x] Build and test new image locally with Azul's `make test`
  - [x] Remove unused dependencies with high or critical CVEs
  - [x] Push commit to GitHub (directly to master branch, no PR needed)
  - [x] GH Action workflow succeeded
  - [x] Image is available on [DockerHub](https://hub.docker.com/repository/docker/ucscgi/azul-elasticsearch) 
- [x] Create Azul PR, connected to this issue, with … 
  - [x] … changes to `requirements*.txt` from open Dependabot PRs, one commit per PR
  - [x] … update to [Python](https://hub.docker.com/_/python/tags) (only patch versions) <sub>or no update available</sub>
  - [x] … update to [Terraform](https://hub.docker.com/r/hashicorp/terraform/tags) (only patch versions) <sub>or no update available</sub>
  - [x] … new [PyCharm image](https://hub.docker.com/repository/docker/ucscgi/azul-pycharm)
  - [x] … new [Elasticsearch image](https://hub.docker.com/repository/docker/ucscgi/azul-elasticsearch)
  - [x] … update to [Docker images](https://hub.docker.com/_/docker/tags) (only minor and patch versions) <sub>or no update available</sub>
  - [x] … update to [GitLab](https://hub.docker.com/r/gitlab/gitlab-ce/tags) & [GitLab runner images](https://hub.docker.com/r/gitlab/gitlab-runner/tags) <sub>or no update available</sub>
  - [x] … update to [ClamAV image](https://hub.docker.com/r/clamav/clamav/tags) <sub>or no update available</sub>
  - [x] … update to [GitLab AMI](https://github.com/DataBiosphere/azul/blob/develop/OPERATOR.rst#updating-the-ami-for-gitlab-instances) <sub>or no update available</sub>
- [ ] ~Delete obsolete image tags from DockerHub (but consider that `prod` may not use the latest image) …~ (this will be tracked as a CL item in the next promotion, it's in the PR template)
  - [ ] ~… for [PyCharm image](https://hub.docker.com/repository/docker/ucscgi/azul-pycharm)~
  - [ ] ~… for [Elasticsearch image](https://hub.docker.com/repository/docker/ucscgi/azul-elasticsearch)~
- [x] Created tickets for any deferred updates to …
  - [x] … to next major or minor Python version <sub>or such ticket already exists</sub>
  - [x] … to next major Docker version <sub>or such ticket already exists</sub>
  - [x] … to next major or minor Terraform version <sub>or such ticket already exists</sub>
- [x] Post vulnerability report for `anvilprod` on this issue

### PR DESCRIPTION
**Pull request by achave11-ucsc**
Connected issue: #5772

### Author

- [x] Target branch is `develop`
- [x] ~Name of PR branch matches `upgrades/yyyy-mm-dd`~ <sub> Name selection preceded this check</sub>
- [x] On ZenHub, PR is connected to the upgrade issue it resolves
- [x] PR title matches `Upgrade dependencies yyyy-mm-dd`
- [x] PR title references the connected issue

### Author (upgrading deployments)

- [x] Added `upgrade` label to PR <sub>or this PR does not require upgrading deployments</sub>

### Author (before every review)

- [x] Rebased PR branch on `develop`, squashed old fixups
- [x] Ran `make requirements_update` <sub>or this PR does not touch requirements*.txt, common.mk, Makefile and Dockerfile</sub>
- [x] Added `R` tag to commit title <sub>or this PR does not touch requirements*.txt</sub>
- [x] Added `reqs` label to PR <sub>or this PR does not touch requirements*.txt</sub>
- [x] Selected `dev.shared` and ran `CI_COMMIT_REF_NAME=develop make -C terraform/shared apply_keep_unused` <sub>or this PR does not change any Docker image versions</sub>
- [x] Selected `anvildev.shared` and ran `CI_COMMIT_REF_NAME=develop make -C terraform/shared apply_keep_unused` <sub>or this PR does not change any Docker image versions</sub>
- [x] Selected `anvilprod.shared` and ran `CI_COMMIT_REF_NAME=develop make -C terraform/shared apply_keep_unused` <sub>or this PR does not change any Docker image versions</sub>
- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>

### System administrator (after approval)

- [x] Actually approved the PR
- [x] Labeled connected issue as `no demo`
- [x] PR title is appropriate as title of merge commit
- [x] Moved ticket to *Approved* column
- [x] PR is assigned to current operator

### Operator (before pushing merge the commit)

- [x] Squashed PR branch and rebased onto `develop`
- [x] Sanity-checked history
- [x] Pushed PR branch to GitHub
- [x] Selected `dev.gitlab` and ran `make -C terraform/gitlab apply` <sub>or this PR does not change the GitLab version</sub>
- [x] Selected `anvildev.gitlab` and ran `make -C terraform/gitlab apply` <sub>or this PR does not change the GitLab version</sub>
- [x] Selected `anvilprod.gitlab` and ran `make -C terraform/gitlab apply` <sub>or this PR does not change the GitLab version</sub>
- [x] Assigned system administrator <sub>or this PR does not change the GitLab version</sub>
- [x] Checked the items in the next section <sub>or this PR changes the GitLab version</sub>

### System administrator

- [x] Background migrations for `dev.gitlab` are complete <sub>or this PR does not change the GitLab version</sub>
- [x] Background migrations for `anvildev.gitlab` are complete <sub>or this PR does not change the GitLab version</sub>
- [x] Background migrations for `anvilprod.gitlab` are complete <sub>or this PR does not change the GitLab version</sub>
- [x] PR is assigned to operator

### Operator (before pushing merge the commit)

- [x] Selected `dev.gitlab` and ran `make -C terraform/gitlab/runner` <sub>or this PR does not change `azul_docker_version`</sub>
- [x] Selected `anvildev.gitlab` and ran `make -C terraform/gitlab/runner` <sub>or this PR does not change `azul_docker_version`</sub>
- [x] Selected `anvilprod.gitlab` and ran `make -C terraform/gitlab/runner` <sub>or this PR does not change `azul_docker_version`</sub>
- [x] Added `sandbox` label
- [x] Pushed PR branch to GitLab `dev`
- [x] Pushed PR branch to GitLab `anvildev`
- [x] Pushed PR branch to GitLab `anvilprod`
- [x] Build passes in `sandbox` deployment
- [x] Build passes in `anvilbox` deployment
- [x] Build passes in `hammerbox` deployment
- [x] Reviewed build logs for anomalies in `sandbox` deployment
- [x] Reviewed build logs for anomalies in `anvilbox` deployment
- [x] Reviewed build logs for anomalies in `hammerbox` deployment
- [x] Title of merge commit starts with title from this PR
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title <sub>but exclude any `p` tags</sub>
- [x] Moved connected issue to Merged column in ZenHub
- [x] Pushed merge commit to GitHub

### Operator (after pushing the merge commit)

- [x] Pushed merge commit to GitLab `dev`
- [x] Pushed merge commit to GitLab `anvildev`
- [x] Pushed merge commit to GitLab `anvilprod`
- [x] Build passes on GitLab `dev`
- [x] Reviewed build logs for anomalies on GitLab `dev`
- [x] Build passes on GitLab `anvildev`
- [x] Reviewed build logs for anomalies on GitLab `anvildev`
- [x] Build passes on GitLab `anvilprod`
- [x] Reviewed build logs for anomalies on GitLab `anvilprod`
- [x] Selected `dev.shared` and ran `make -C terraform/shared apply` <sub>or this PR does not change any Docker image versions</sub>
- [x] Selected `anvildev.shared` and ran `make -C terraform/shared apply` <sub>or this PR does not change any Docker image versions</sub>
- [x] Selected `anvilprod.shared` and ran `make -C terraform/shared apply` <sub>or this PR does not change any Docker image versions</sub>
- [x] Deleted PR branch from GitHub
- [x] Deleted PR branch from GitLab `dev`
- [x] Deleted PR branch from GitLab `anvildev`
- [x] Deleted PR branch from GitLab `anvilprod`

### Operator

- [x] Ran `script/export_inspector_findings.py` against `anvilprod`, imported results to [Google Sheet](https://docs.google.com/spreadsheets/d/1RWF7g5wRKWPGovLw4jpJGX_XMi8aWLXLOvvE5rxqgH8) and posted screenshot of relevant<sup>1</sup> findings as [a comment on the connected issue.](https://github.com/DataBiosphere/azul/issues/5772#issuecomment-1865224369)
- [x] PR is assigned to system administrator

<sup>1</sup>A relevant finding is a high or critical vulnerability in an image
that is used within the security boundary. Images not used within the boundary
are tracked in `azul.docker_images` under a key starting with `_`.

### System administrator

- [x] No currently reported vulnerability requires immediate attention
- [x] PR is assigned to no one

## Shorthand for review comments

- `L` line is too long
- `W` line wrapping is wrong
- `Q` bad quotes
- `F` other formatting problem

### CODE REVIEW

**Comment by achave11-ucsc** [comment]
The checklist of upgrades issue #5772 has been updated to reflect the changes introduced in this PR.

**Review by dsotirho-ucsc — DISMISSED** [review]
Nit: My preference would be to use the standard PR template for these `upgrade dependencies` PRs, unless/until we create a new template specific for these upgrades. 

With that said, it appears the current PR checklist covers all the key items, so approved.

File: README.md [hunk 1]
```diff
@@ -1034,7 +1034,7 @@ found in [`AzulTestCase`](test/azul_test_case.py) and commit the modifications.
 
 ## Setting up the Azul build prerequisites on macOS 12 (Monterey)
 
-The steps below are examplary for Python 3.11.6. Replace `3.11.6` with the 
+The steps below are examplary for Python 3.11.7. Replace `3.11.7` with the 
 version listed in [environment.boot](environment.boot).
```
Comment (hannes-ucsc) [inline README.md]: Please revert changes to this file. If we were to update it in every upgrade PR, the very sentence you're modifying here would be redundant.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
I think we should merge this PR before we merge my https://github.com/DataBiosphere/azul/pull/5806. One thing you should steal from that PR though is the "split" deployment of the `shared` component so that personal deployments don't break while this PR is being merged. Look for `apply_keep_unused` in `.github/PULL_REQUEST_TEMPLATE/upgrade.md`.

**Review by hannes-ucsc — DISMISSED** [review]

**Comment by achave11-ucsc** [comment]
@hannes-ucsc, perhaps we could add the following patch, make it less of a 'figure it out' thing for deploying the GitLab component. That's because we're deploying from the feature branch, as the CL currently stands. Or this might be a new improvisations issue.

```diff
Subject: [PATCH] fixup! Fix: CL item for running IT in personal deployments missing from upgrade PR template
---
Index: .github/PULL_REQUEST_TEMPLATE/upgrade.md
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/.github/PULL_REQUEST_TEMPLATE/upgrade.md b/.github/PULL_REQUEST_TEMPLATE/upgrade.md
--- a/.github/PULL_REQUEST_TEMPLATE/upgrade.md	(revision 5bb85c1fc013d5303bb028e0b38b09ca389d791e)
+++ b/.github/PULL_REQUEST_TEMPLATE/upgrade.md	(date 1703050587410)
@@ -48,9 +48,9 @@
 - [ ] Squashed PR branch and rebased onto `develop`
 - [ ] Sanity-checked history
 - [ ] Pushed PR branch to GitHub
-- [ ] Selected `dev.gitlab` and ran `make -C terraform/gitlab apply` <sub>or this PR does not change the GitLab version</sub>
-- [ ] Selected `anvildev.gitlab` and ran `make -C terraform/gitlab apply` <sub>or this PR does not change the GitLab version</sub>
-- [ ] Selected `anvilprod.gitlab` and ran `make -C terraform/gitlab apply` <sub>or this PR does not change the GitLab version</sub>
+- [ ] Selected `CI_COMMIT_REF_NAME=develop dev.gitlab` and ran `make -C terraform/gitlab apply` <sub>or this PR does not change the GitLab version</sub>
+- [ ] Selected `CI_COMMIT_REF_NAME=develop anvildev.gitlab` and ran `make -C terraform/gitlab apply` <sub>or this PR does not change the GitLab version</sub>
+- [ ] Selected `CI_COMMIT_REF_NAME=develop anvilprod.gitlab` and ran `make -C terraform/gitlab apply` <sub>or this PR does not change the GitLab version</sub>
 - [ ] Assigned system administrator <sub>or this PR does not change the GitLab version</sub>
 - [ ] Checked the items in the next section <sub>or this PR changes the GitLab version</sub>
 
Index: .github/pull_request_template.md.template.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/.github/pull_request_template.md.template.py b/.github/pull_request_template.md.template.py
--- a/.github/pull_request_template.md.template.py	(revision 5bb85c1fc013d5303bb028e0b38b09ca389d791e)
+++ b/.github/pull_request_template.md.template.py	(date 1703050581046)
@@ -548,7 +548,8 @@
                 *[
                     {
                         'type': 'cli',
-                        'content': f'Selected `{deployment}.gitlab` and '
+                        'content': f'Selected `{"CI_COMMIT_REF_NAME=develop " if t == T.upgrade else ""}'
+                                   f'{deployment}.gitlab` and '
                                    f'ran `make -C terraform/gitlab apply`',
                         'alt': 'or this PR does not change the GitLab version'
                     }

```

**Review by hannes-ucsc — APPROVED** [review]

**Comment by achave11-ucsc** [comment]
Screen Shot 2023-12-20 at 12 38 44 PM

**Comment by hannes-ucsc** [comment]
@achave11-ucsc, skipping the sandbox builds after the `botocore-stubs` dependency bump  caused [an issue](https://gitlab.prod.anvil.gi.ucsc.edu/ucsc/azul/-/jobs/18283) with the next PR.

**Pull request merged**